diff --git a/src/src/smtp_in.c b/src/src/smtp_in.c
index 2fc6fd20e..bb4288ef4 100644
--- a/src/src/smtp_in.c
+++ b/src/src/smtp_in.c
@@ -3625,7 +3625,7 @@ log_close_event(US"by QUIT");
 /* Pause, hoping client will FIN first so that they get the TIME_WAIT.
 The socket should become readble (though with no data) */
 
-(void) poll_one_fd(fileno(smtp_in), POLLIN, 200);
+if (smtp_in) (void) poll_one_fd(fileno(smtp_in), POLLIN, 200);
 #endif	/*!SERVERSIDE_CLOSE_NOWAIT*/
 }
 
