diff --git a/src/src/functions.h b/src/src/functions.h
index bf72615f8..45ccfc21a 100644
--- a/src/src/functions.h
+++ b/src/src/functions.h
@@ -1073,7 +1073,8 @@ static inline gstring * string_cat(gstring * g, const uschar * s)
 static inline gstring *
 string_cat(gstring * g, const uschar * s)
 {
-return string_catn(g, s, Ustrlen(s));
+if (!s) return g;
+else return string_catn(g, s, Ustrlen(s));
 }
 
 
